test-interface/src/main/scala/org/scalajs/testinterface/TestUtils.scala

```diff
@@ -1,24 +1,145 @@
 package org.scalajs.testinterface
 
 import scala.scalajs.js
+import scala.scalajs.reflect._
 
 object TestUtils {
 
-  def newInstance(name: String, loader: ClassLoader)(args: Seq[AnyRef]): AnyRef = {
-    val ctor = deepSelect(namespace(loader), name)
-    js.Dynamic.newInstance(ctor)(args.asInstanceOf[Seq[js.Any]]: _*)
+  /** Instantiates the class given by its fully qualified name.
+   *
+   *  The target class must either
+   *
+   *  - extend a class or trait annotated with
+   *    [[scala.scalajs.reflect.annotation.EnableReflectiveInstantiation]], or
+   *  - be exported under its fully qualified name.
+   *
+   *  In the former case, a best-effort is made to find the appropriate
+   *  constructor overload.
+   *
+   *  Prefer using the other overload of `newInstance` for new code, which
+   *  allows you to explicitly specify the constructor overload to use.
+   */
+  @deprecated(
+      "Use the overload with explicit formal constructor parameter types.",
+      "0.6.15")
+  def newInstance(name: String, loader: ClassLoader)(
+      args: Seq[AnyRef]): AnyRef = {
+    Reflect.lookupInstantiatableClass(name).fold[AnyRef] {
+      val ctor = deepSelect(namespace(loader), name)
+      js.Dynamic.newInstance(ctor)(args.asInstanceOf[Seq[js.Any]]: _*)
+    } { clazz =>
+      val ctor = guessAppropriateCtor(clazz.declaredConstructors, args)
+      ctor.newInstance(args: _*).asInstanceOf[AnyRef]
```

Comment:
In fact, [as we can see here](https://github.com/sjrd/scala-js/blob/v0.6.14/stubs/src/main/scala/org/scalajs/testinterface/TestUtils.scala#L28-L31), the JVM version of `newInstance` in master doesn't even try to resolve overloaded constructors. If there is more than one public constructor, I think it would be fair to bail here as well, instead of this `guessAppropriateCtor` thing which is super fragile.